proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,90 @@
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+* Impacted Projects: Clang
+
+## Introduction
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors) 
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+ we need to be able to scalarize datastructures, call instructions, and vector 
+ operations like mathamatical, logical and bitcasts. The goal of this proposal 
+ is to present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate
+ valid DXIL.
+
+ ## Background
+ In DXC Scalarization and SROA are merged in unfortnate ways. Most of that code
+ can be found here: [`SROA_Parameter_HLSL` Module pass in `ScalarReplAggregatesHLSL.cpp`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L4263). 
+The specific instructions DXC operates on are `CallInst`, `LoadInst`, 
+`StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, and 
+`MemIntrinsic`. Not covered in this proposal is the handling of `Allocas`.
+That will be addressed in a later proposal covering DXIL Legalization.
+
+## Proposal
+Our goal should be legal DIXL with behavioral similarity to DXC. To that end it
+ makes the most sense to use the Scalarizer pass.
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+The `scalarizer` pass with the `-scalarize-load-store` and 
+`-scalarize-variable-insert-extract` flags  will handle the `BitCastInst`, 
+`CallInst`, `LoadInst` and `StoreInst`, and `GetElementPtrInst` cases. The 
+`AddrSpaceCastInst` case isn't relevant for upstream LLVM.
+
+`MemIntrinsic` is the odd case out. We don't currently emit memcpy for
+the same cases we do in DXC.
+In many cases operations like memcpy and memset seem to happen in global scope.
+See Example:
+- https://hlsl.godbolt.org/z/8vx1Yee7b
+
+In my observation these cases get converted into cbuffer. Since we don't have
+cbuffer support yet we can likely hold off on this cases.
+
+
+The team debated when how early or late this pass should run and three ways to 
+onboard this pass. There was also a discussion on  the scalarizer passes only 
+working on llvm intrinsics and how to allow for DirectX Intrinsics.
+
+### Proposal: When to run pass
+The team determined it was best for this pass to run in the DirectX backend.
+That allows The DirectX backend to be agnostic to the frontend.
+The team also determined the pass should run as late as possible. This has two
+benefits. First code size. if scalarization happens too early things like the 
+`-combiner-alias-analysis` pass limits are reached. More optimizations will be
+possible as analysis passes will benefit from less IR. Second, scalarization 
```

Comment:
> More optimizations will be possible as analysis passes will benefit from less IR.

Is this saying that the analysis passes will be able to perform better analysis if there is less IR?  Or do the pass limits mentioned previously mean that some passes are disabled?

Reply:
There are some passes that have hard limits on code size `combiner-alias-analysis` was one of them. In fact, load store scalarization isn't on by default because it can affect that particular pass.